### ISSUE TITLE
Write major modes bindings

### ISSUE THREAD

**Issue by MarcoIeni** [VSpaceCode/VSpaceCode#98]
**Is your feature request related to a problem? Please describe.**

As I said [here](https://github.com/VSpaceCode/vscode-which-key/issues/11#issuecomment-671416989):

> In spacemacs you have major-modes, like "python", "markdown", and so on. So for example if you open a `.md` file, spacemacs will use Markdown mode in that buffer.
>
> When you are using a mode, key bindings under `<spc> m` or under `,` will be unique to that mode.
>
> This means that when you are editing a markdown, `<spc> m b` could be assigned to "make bold", while in python `<spc> m b` could be assigned to "toggle breakpoint" (I am just inventing).

**Describe the solution you'd like**

In [vscode-which-key/#14](https://github.com/VSpaceCode/vscode-which-key/pull/14) we have enabled the possibility to have major mode key bindings. Now it's time to write them!

These are some of the [most important](https://madnight.github.io/githut/#/pull_requests/2020/2) programming languages where a major mode is missing:

- [x] javascript
- [x] python
- [ ] java - eric on it
- [x] go
- [x] cpp
- [x] ObjectPascal
- [x] typescript
- [x] php
- [x] c#
- [ ] c
- [ ] shell
- [ ] scala
- [x] rust
- [x] clojure
- [x] ruby
- [x] F#
- [ ] R
- [ ] jsx, tsx

markup languages:

- [x] markdown
- [ ] org mode

If you want to start working on a major mode that is (or isn't) listed here, please leave a comment.
In this way we don't overlap with each other.

**Additional context**
- In 23b5b75 you can find an example of a markdown major mode binding.
- [top vs code extensions](https://marketplace.visualstudio.com/search?target=VSCode&category=All%20categories&sortBy=Installs)

**Comment by stevenguh** [comment]
Took this out the v0.8.0 because this might take a bit more brewing.

**Comment by macintacos** [comment]
I was thinking about tackling the Markdown major mode, however I have to ask - can I assume that [Markdown All in One](https://marketplace.visualstudio.com/items?itemName=yzhang.markdown-all-in-one) would be a "dependency"? It's by far the most popular Markdown extension, but I'm not sure if you want to limit the amount of extensions that VSpaceCode relies on.

**Comment by stevenguh** [comment]
Thank you for taking an interest in implementing major modes :)

Since not everyone is going to need/want to install the extension required by a specific default major mode, our thought is to not bundle the extension (required by major modes) and require manual installation for now. Follow https://github.com/VSpaceCode/VSpaceCode/issues/84#issuecomment-686802124 for the context.

Since the engineering for a major mode bundle system is not here yet (refer to https://github.com/VSpaceCode/VSpaceCode/issues/84#issuecomment-686802124), we plan to do something simple here. We will ship a default major bindings that require user to manually install the missing extension if needed.

In short, I am open to have a default markdown bindings that require manually installation of an extension. @MarcoIeni What do you think?

**Comment by MarcoIeni** [comment]
Macintacos, you can assume that the extension is there.

At the moment we need to document that it requires manual installation.
In the future we might decide to boundle it.

**Comment by stevenguh** [comment]
> For this issue it doesn't matter, right? We might decide this once the markdown mode is "complete"

I was thinking markdown is probably a little bit special, so we can bundle the extension needed directly to this extension. But thinking it more, it's good to not bundle any extensions required by major mode for now to be consistent (until we have the system mentioned perviously)

**Comment by macintacos** [comment]
@MarcoIeni @stevenguh awesome, I'll get started on a proper PR soon, I'll link it to this issue when I open it.

**Comment by stevenguh** [comment]
Released v0.8.2 that contains the new markdown mode. Thank you @macintacos for contributing :)

**Comment by wneville** [comment]
Hello, was very pleased to stumble across this project. I'd love to help out and take a stab at a major mode binding.

It's not on the list, but one that I was wishing for today is `HTTP` support for the [`REST Client` extension](https://github.com/Huachao/vscode-restclient).

The [Spacemacs bindings](https://www.spacemacs.org/layers/+tools/restclient/README.html) for the relevant Emacs package are pretty sparse, and some of them aren't relevant. The REST Client extension in VSCode pulls focus away from the editor on sending a request and the HTTP response appears to be in a [readonly webview](https://github.com/Huachao/vscode-restclient/blob/master/src/views/httpResponseWebview.ts) by default, which prevents invoking followup commands with keystrokes (as far as I'm currently aware).

Regardless, there are a number of commands that would be useful to be exposed via `<spc> m` when in an HTTP language buffer - see screenshot below. Would this one make sense to add to the list of major modes?

![image](https://user-images.githubusercontent.com/3227558/95634846-d90f7200-0a58-11eb-82ff-fbc522d9aa5b.png)

**Comment by stevenguh** [comment]
Thank you for taking an interest in contributing. Although, this is not in the list, I am open to it. As far as to activate the menu from the readonly WebViews, we will have to target it as a shortcut with keybindings.json and its corresponding context like https://github.com/VSpaceCode/VSpaceCode/blob/3cbec1541f730ecb9ceff8af5ecd2a9e6a478369/src/configuration/keybindings.jsonc#L3-L7

You can use inspect context command with and developer prompt to see what can be used to target. 
<img width="386" alt="image" src="https://user-images.githubusercontent.com/3455662/95649057-7c2fae00-0a90-11eb-9e10-52e76a58eef9.png">
e.g. `activeEditor == 'WebviewEditor'` might be one of the when/context condition you can used to activate on that readonly view.

**Comment by MarcoIeni** [comment]
It's fine for me to support this extension, even if I haven't used it in spacemacs.
Anyway, maybe it would be better to not limit this key bindings to the `html` major mode, because it could be useful also in all the other modes I think.

Spacemacs has `<spc> a` which stands for "applications". We could have something like `<spc> a c` (for client) or `<spc> a r` (for rest client).
What do you think?

Of course the approach is the same as major modes, i.e. we don't bundle this extension, user have to install it manually.

> The Spacemacs bindings for the relevant Emacs package are pretty sparse, and some of them aren't relevant. 

If users are used to the spacemacs key bindings for this functionality and if they make sense for vs code we can copy them here, otherwise we can ignore them.

**Comment by stevenguh** [comment]
> Anyway, maybe it would be better to not limit this key bindings to the html major mode, because it could be useful also in all the other modes I think.

I think that extension uses `http` language file, but so I think it make sense to be in major mode. The second part is the extension creates a read-only WebViews editor. We have to find the right context/when to target specific that window, and we might not able to do that; therefore, I am fine with the additional of `http` language mode major mode only.

> Spacemacs has <spc> a which stands for "applications". We could have something like <spc> a c (for client) or <spc> a r (for rest client).
What do you think?

I will be a little bit hesitant to bundle commands that required other extensions outside of major mode for now.

> Anyway, can you create another issue, so we move the conversation there?

+1

**Comment by Muscaw** [comment]
Hey, I wanted to know if I could contribute to the extension by working on the Python major mode.

**Comment by MarcoIeni** [comment]
Hi Muscaw, sure! I will reserve it to you in the first comment :)

**Comment by jr0cket** [comment]
@MarcoIeni I am interested in adding Major mode key bindings for the Clojure programming language, using the popular Calva plugin for VS Code.

It seems to test changes I edit the `package.json` file that is extracted to my `$HOME/.vscode/extensions/vspacecode.vspacecode-0.8.5 directory`, restarting VS Code for changes to take effect.  

I just copied some of the go configuration and changed it to use `languageId:clojure` and this does give a menu under `SPC m` and `,` when I have a Clojure file open.

**Comment by stevenguh** [comment]
Feel free to open an issue to track it if you want to work on it :)

> It seems to test changes I edit the package.json file that is extracted to my
> $HOME/.vscode/extensions/vspacecode.vspacecode-0.8.5 directory, restarting VS Code for changes to take effect.

That's very interesting way to test it. You will be modifying the built-in default from the extension directly I think. You can follow the [CONTRIBUTING](https://github.com/VSpaceCode/VSpaceCode/blob/master/CONTRIBUTING.md) to run a test instance with the updated default config.

**Comment by MarcoIeni** [comment]
Thanks jr0cket! I just updated the first comment/

**Comment by mayanez** [comment]
I'd be happy to try my hand at the C++ major mode.

**Comment by stevenguh** [comment]
> I'd be happy to try my hand at the C++ major mode.

Awesome! Please open an issue to track that work :)

**Comment by nodeselector** [comment]
Working on a javascript implementation.

**Comment by cimentadaj** [comment]
Any interest in an R major mode?

**Comment by basus** [comment]
I'm interested in helping add an OCaml major mode. Since the [VS Code OCaml Platform plugin](https://github.com/ocamllabs/vscode-ocaml-platform/issues?q=build+dune+projects) is not very feature-rich at the moment, I would like to develop it locally for a while, until I can get some changes into that plugin. Is there some way to experiment with a major mode locally without putting it into `packages.json`? I tried something like this, but it doesn't seem to work:
```
"vspacecode.bindingOverrides": [
        {
            "keys": "m.languageId:ocaml",
            "name": "OCaml",
            "type": "bindings",
            "bindings": [... ]
        } ]
```

**Comment by The-Compiler** [comment]
@basus Does `"keys": ["m", "languageId:ocaml"],` work? Here's an example of what works for me:

https://github.com/The-Compiler/dotfiles/blob/f1778cc2565c1cfc975b601b8e398012f5b26b24/vscode/settings.jsonc#L621-L642

**Comment by basus** [comment]
Looks like @The-Compiler's suggestion worked. Thanks!

Now the issue is, I want the same keybindings to work for both `languageId.ocaml` and `languageId.ocaml.interface`, but I can't tell if there is a straightforward way to do that?

**Comment by The-Compiler** [comment]
You'll need to duplicate them for now, see https://github.com/VSpaceCode/vscode-which-key/issues/42

**Comment by erichonorez** [comment]
Hi,

I'm currently working on a Java major mode [here](https://github.com/erichonorez/VSpaceCode/tree/java-major-mode).
I still need to do a some of tests and once ready I'll do a PR.

**Comment by e-dorigatti** [comment]
Would it be possible to implement some default keybindings that work for any major mode? For example, many "Go to" commands are already implemented in VSCode (e.g., `editor.action.revealDefinition` for "go to definition"), and I would expect to be able to use these in the same way regardless of the active major mode, *especially* for major modes that are not implemented yet.

This would greatly reduce the burden to implement and maintain major modes for each and every language, and having this basic functionality working out of the box would greatly improve the default user experience. Indeed, I am writing this comment because I happened to try this extension on a TypeScript project, which is [not yet](https://github.com/VSpaceCode/VSpaceCode/issues/272) implemented.

(sorry if this was already asked, I jus stumbled upon this extension and it looks great btw)

**Comment by stevenguh** [comment]
> Would it be possible to implement some default keybindings that work for any major mode? For example, many "Go to" commands are already implemented in VSCode (e.g., `editor.action.revealDefinition` for "go to definition"), and I would expect to be able to use these in the same way (e.g., `,gd`) regardless of the active major mode, _especially_ for languages that do not have a major mode implemented.
> 
> This would greatly reduce the burden to implement and maintain major modes for each and every language, and having this basic functionality working out of the box would greatly improve the default user experience. Indeed, I am writing this comment because I happened to try this extension on a TypeScript project, which is [not yet](https://github.com/VSpaceCode/VSpaceCode/issues/272) implemented.
> 
> (sorry if this was already asked, I jus stumbled upon this extension and it looks great btw)

This is a reasonable ask. Especially, lots of major mode's command are basically just standard vscode command as long as there's LSP for the language. 

Unfortunately. I don't really have the bandwidth to look at this too closely. Maybe in a few months :'(

**Comment by 0xadeeb** [comment]
Is there any work going on to write major mode for C lang? If not I'd like to start working on it.

**Comment by MarcoIeni** [comment]
No ongoing work as far as I'm aware. Go for it 👍
